### DISCUSSION TITLE
Compiling for ESP32

### DISCUSSION THREAD

**Discussion by h-milz** [georgeweigt/eigenmath#27] [Ideas]
Hi, I am exploring the possibility of using eigenmath for a calculator based on an ESP32-S3 with LVGL in Arduino or ESP-IDF mode. Someone created an eigenmath fork two years ago and made a Micropython module (largely unmaintained now) from ancient versions of eigenmath (from 2021) and LVGL v7 I think, but I don't feel like reverse engineering what they did in detail (the diff against the recent eigenmath code is a mess) and rather start from scratch and without Micropython.

In general, the recent eigenmath code compiles as-is with the xtensa compiler albeit with a couple of warnings that are xtensa specific and can be ignored for now. Reading your source code, I think I understand by and large how things work, but what I did not understand so far is the role of the following configuration items:

```
#define STACKSIZE 100000 // evaluation stack
#define FRAMESIZE 10000
#define BLOCKSIZE 10000
#define MAXBLOCKS 2000
#define BUCKETSIZE 100
#define STRBUFLEN 1000
#define MAXDIM 24
```
What are FRAMESIZE, BLOCKSIZE, MAXBLOCKS and BUCKETSIZE for? Can you please shed some light on it? I understand these defines determine the size of various items which is important on a RAM constrained machine (albeit with 8 MB octal SPIRAM). How would you size these defines for this architecture? 

Also, where would I hook the LVGL drawing primitives in? What I think I understood is that xcode/eval_draw.c should be reusable more or less as-is but in emit.c, the actual drawing takes place (as in "the drawing is prepared and then sent off to the MacOS library").  In my version of emit_box() for example I could draw a box using LVGL primitives (lines), etc. Does that sound reasonably plausible?

Thank you!

**Comment by georgeweigt** [comment]
frame[FRAMESIZE] was a stack for local variables, now removed.
Symbol names are hashed on the first letter with space for BUCKETSIZE names per letter.
Memory is malloc'd as needed in blocks of BLOCKSIZE * sizeof (struct atom), up to MAXBLOCKS.
For a 64-bit machine sizeof (struct atom) is 24 bytes so a BLOCKSIZE of 10,000 results in a malloc size of 240,000 bytes.
Changing MAXBLOCKS to 10 would result in 2,400,000 bytes malloc'd max.
The command "status" prints out statistics for memory usage which is useful for tuning.
draw_char(), draw_stroke() and draw_point() in draw_display.c are the drawing primitives that would need conversion to LVGL.
If ESP32 supports a browser then eigenmath.js is another option for providing a calculator.
Best regards,
gw

**Comment by h-milz** [comment]
Ah, I got it to compile and run in Arduino mode with very few changes, just a few ifdefs and allocating the buffers in PSRAM, and on the 8MB PSRAM device the default values work fine. On such a machine I don't expect much load anyway, as opposed to some of your examples. 

So I think I will make a new dir parallel to js and xcode, and begin experimenting there, including a simple micropython binding for run(). Should then work as a generic micropython module as well as a C module for ESP32 in whatever IDE. Should in fact run on any microcontroller with enough PSRAM, like the Teensy4.1. 

JS is not an option - afaik there is no JS capable browser for any of these. Resources ... 

Thank you for this great piece of work!
